File: safety_node/scripts/safety_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

import numpy as np
# TODO: include needed ROS msg type headers and libraries
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import Odometry
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive


class SafetyNode(Node):
    """
    The class that handles emergency braking.
    """
    def __init__(self):
        super().__init__('safety_node')
        """
        One publisher should publish to the /drive topic with a AckermannDriveStamped drive message.

        You should also subscribe to the /scan topic to get the LaserScan messages and
        the /ego_racecar/odom topic to get the current speed of the vehicle.

        The subscribers should use the provided odom_callback and scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        """
        self.speed = 0.
        # create ROS subscribers and publishers.
        self.drive_publisher_ = self.create_publisher(AckermannDriveStamped, "drive", 10)
        self.odom_subscriber_ = self.create_subscription(Odometry, "ego_racecar/odom", self.odom_callback, 10)
        self.scan_subscriber_ = self.create_subscription(LaserScan, "scan", self.scan_callback, 10)
        self.get_logger().info("Safety node has been started")
        

    def odom_callback(self, odom_msg):
        # update current speed
        self.speed = odom_msg.twist.twist.linear.x

    def scan_callback(self, scan_msg):
        # calculate TTC
        # get the array of scan angels
        scan_angles = np.arange(start=scan_msg.angle_min, stop=scan_msg.angle_max, step=scan_msg.angle_increment, dtype=np.float32)
        scan_angles = scan_angles[:len(scan_msg.ranges)]
        # get the array of range rates
        range_rates = self.speed * np.cos(scan_angles)
        # make the rate zero when non-positive
        range_rates = np.maximum(range_rates,0)
        # get the TTC array
        with np.errstate(divide='ignore', invalid='ignore'):
            TTC_array = np.where(np.isnan(scan_msg.ranges), float('inf'), np.where(range_rates!=0, scan_msg.ranges/range_rates, float('inf')))
        
        # publish command to brake
        drive_msg = AckermannDriveStamped ()
        # TTC Condition
        TTC = min(TTC_array)
        if (TTC<=1.5):
            drive_msg.drive.speed = 0.0
            self.drive_publisher_.publish(drive_msg)
            self.get_logger().info("AEB Activated with TCC: " + str(TTC))
        else:
            pass

def main(args=None):
    rclpy.init(args=args)
    safety_node = SafetyNode()
    rclpy.spin(safety_node)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    safety_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: safety_node/src/safety_node.cpp
#include "rclcpp/rclcpp.hpp"
/// include needed ROS msg type headers and libraries
#include "sensor_msgs/msg/laser_scan.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"

#include <vector>
#include <algorithm>
#include <cmath>


class Safety : public rclcpp::Node {
// The class that handles emergency braking

public:
    Safety() : Node("safety_node")
    {
        /*
        You should also subscribe to the /scan topic to get the
        sensor_msgs/LaserScan messages and the /ego_racecar/odom topic to get
        the nav_msgs/Odometry messages

        The subscribers should use the provided odom_callback and 
        scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        */

        /// create ROS subscribers and publishers
        drive_publisher_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("drive", 10);
        odom_subscriber_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "ego_racecar/odom", 10, std::bind(&Safety::drive_callback, this, std::placeholders::_1));
        scan_subscriber_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "scan", 10, std::bind(&Safety::scan_callback, this, std::placeholders::_1));
        RCLCPP_INFO(this->get_logger(), "Safety Node has been started");
    }

private:
    // initiate some global values
    double speed = 0.0;

    // double time_prev = 0.0;
    // std::vector<float> ranges_prev;

    /// create ROS subscribers and publishers
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_publisher_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_subscriber_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_subscriber_;

    void drive_callback(const nav_msgs::msg::Odometry::ConstSharedPtr msg)
    {
        /// update current speed
        speed = msg->twist.twist.linear.x;
    }

    void scan_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) 
    {
        /// Calculate TTC
        // Get the array of scan angles
        float angle_min = scan_msg->angle_min;
        // float angle_max = scan_msg->angle_max;
        float angle_increment = scan_msg->angle_increment;
        int num_ranges = scan_msg->ranges.size();
        
        std::vector<float> scan_angles(num_ranges);
        for (int i = 0; i < num_ranges; ++i) 
        {
            scan_angles[i] = angle_min + i * angle_increment;
        }

        // Get the array of range rates
        std::vector<float> range_rates(num_ranges);
        for (int i = 0; i < num_ranges; ++i) 
        {
            range_rates[i] = std::max(speed * std::cos(scan_angles[i]), 0.0);
        }

        // Get the TTC array
        std::vector<float> TTC_array(num_ranges);
        for (int i = 0; i < num_ranges; ++i) {
            if (std::isnan(scan_msg->ranges[i])) {
                TTC_array[i] = std::numeric_limits<float>::infinity();
            } else {
                TTC_array[i] = (range_rates[i] != 0) ? scan_msg->ranges[i] / range_rates[i] : std::numeric_limits<float>::infinity();
            }
        }

        /// publish drive/brake message
        float min_value = TTC_array[0];
        int min_index = 0;

        for (int i = 1; i < num_ranges; i++) {
            if (TTC_array[i] < min_value) {
                min_value = TTC_array[i];
                min_index = i;
            }
        }

        if (min_value < 1.3)
        {
            auto drive_msg = ackermann_msgs::msg::AckermannDriveStamped();
            drive_msg.drive.speed = 0.0;
            drive_publisher_ ->publish(drive_msg);
            float AEB_angle = (angle_min + min_index * (angle_increment)) * (180.0 / M_PI);
            RCLCPP_INFO(this->get_logger(), "AEB activated with TTC:  %f at the angle %f", min_value, AEB_angle);
        }
    }
};


int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Safety>());
    rclcpp::shutdown();
    return 0;
}